File: iosApp/iosApp/Features/Calendar/CalendarViewModel.swift
//
//  CalendarViewModel.swift
//  BDE
//
//  Created by Nathan Fallet on 19/05/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

class CalendarViewModel: ObservableObject {
    
    @Published var day = Date() { didSet { updateTodayEvents() }}
    @Published var events = [CalendarEvent]() { didSet { updateTodayEvents() }}
    @Published var todayEvents = [CalendarEvent]()
    
    func onAppear(token: String?) {
        AnalyticsService.shared.log(.screenView(screenName: "calendar", screenClass: "CalendarView"))
        
        fetchData(token: token, reload: false)
    }
    
    func fetchData(token: String?, reload: Bool) {
        Task {
            let events = try await CacheService.shared.getEvents(token: token, reload: reload)
            DispatchQueue.main.async {
                self.events = events
            }
            if !reload {
                let events = try await CacheService.shared.getEvents(token: token, reload: true)
                DispatchQueue.main.async {
                    self.events = events
                }
            }
        }
    }
    
    func updateTodayEvents() {
        let components = Calendar.current.dateComponents([.day, .month, .year], from: day)
        todayEvents = events.filter({ event in
            Calendar.current.dateComponents([.day, .month, .year], from: event.start?.asDate ?? Date()) == components ||
            Calendar.current.dateComponents([.day, .month, .year], from: event.end?.asDate ?? Date()) == components
        })
    }
    
    func previous() {
        day = Calendar.current.date(byAdding: .day, value: -1, to: day) ?? Date()
    }
    
    func next() {
        day = Calendar.current.date(byAdding: .day, value: 1, to: day) ?? Date()
    }
    
}

File: androidApp/src/main/java/me/nathanfallet/bdeensisa/features/calendar/CalendarViewModel.kt
package me.nathanfallet.bdeensisa.features.calendar

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.analytics.ktx.logEvent
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.launch
import kotlinx.datetime.Clock
import kotlinx.datetime.Instant
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime
import me.nathanfallet.bdeensisa.database.DatabaseDriverFactory
import me.nathanfallet.bdeensisa.extensions.SharedCacheService
import me.nathanfallet.bdeensisa.models.CalendarEvent
import kotlin.time.Duration.Companion.days

class CalendarViewModel(
    application: Application,
    token: String?
) : AndroidViewModel(application) {

    // Properties

    private val day = MutableLiveData(Clock.System.now())
    private val events = MutableLiveData<List<CalendarEvent>>()
    private val todayEvents = MutableLiveData<List<CalendarEvent>>()

    // Getters

    fun getDay(): LiveData<Instant> {
        return day
    }

    fun getTodayEvents(): LiveData<List<CalendarEvent>> {
        return todayEvents
    }

    // Methods

    init {
        Firebase.analytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW) {
            param(FirebaseAnalytics.Param.SCREEN_NAME, "calendar")
            param(FirebaseAnalytics.Param.SCREEN_CLASS, "CalendarView")
        }

        fetchData(token, false)
    }

    fun fetchData(token: String?, reload: Boolean) {
        viewModelScope.launch {
            try {
                SharedCacheService.getInstance(DatabaseDriverFactory(getApplication()))
                    .getEvents(token, reload)
                    .let {
                        events.value = it
                        updateTodayEvents()
                    }
                if (!reload) {
                    SharedCacheService.getInstance(DatabaseDriverFactory(getApplication()))
                        .getEvents(token, true)
                        .let {
                            events.value = it
                            updateTodayEvents()
                        }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun updateTodayEvents() {
        val todayLocal = day.value?.toLocalDateTime(TimeZone.currentSystemDefault())?.date
        todayEvents.value = events.value?.filter {
            it.start?.toLocalDateTime(TimeZone.currentSystemDefault())?.date == todayLocal ||
                    it.end?.toLocalDateTime(TimeZone.currentSystemDefault())?.date == todayLocal
        }
    }

    fun previous() {
        day.value = day.value?.minus(1.days) ?: Clock.System.now()
        updateTodayEvents()
    }

    fun next() {
        day.value = day.value?.plus(1.days) ?: Clock.System.now()
        updateTodayEvents()
    }

}